extensions/amp-analytics/0.1/visibility-manager-for-mapp.js

```diff
@@ -0,0 +1,169 @@
+    this.visibilityInterface_ = visibilityInterface;
+
+    /** @const @private {boolean} */
+    this.backgroundedAtStart_ = !this.viewer_.isVisible();
+
+    /** @private {?../../../src/layout-rect.LayoutRectDef} */
+    this.intersectionRect_ = null;
+
+    /** @private {boolean} */
+    this.disposed_ = false;
+
+    // Initate the listener
+    this.listenToVisibilityChange_();
+  }
+
+  /** @override */
+  dispose() {
+    super.dispose();
+    this.disposed_ = true;
+  }
+
+  /** @override */
+  getStartTime() {
+    // viewer.getFirstVisibleTime depend on the visibilitychange API and
+    // document['hidden']
+    // Expect the viewer is always visible in webview
+    return dev().assertNumber(this.viewer_.getFirstVisibleTime());
+  }
+
+  /** @override */
+  isBackgrounded() {
+    // Listens to visibilitychange event, in theory this never fires
+    return !this.viewer_.isVisible();
+  }
+
+  /** @override */
+  isBackgroundedAtStart() {
+    // Return the first visible state. In theory this is always true in mApp
+    return this.backgroundedAtStart_;
+  }
+
+
+  /** @override */
+  getRootMinOpacity() {
+    // Copied the implementation from VisibilityManagerForDoc,
+    // doesn't count iframe opacity
+    const root = this.ampdoc.getRootNode();
+    const rootElement = dev().assertElement(
+        root.documentElement || root.body || root);
+    return getMinOpacity(rootElement);
+  }
+
+  /** @override */
+  listenElement() {
+    // #listenElement not supported in mApp
+    devAssert(false, '%s: element level visibility not supported, ' +
+        'getElementIntersectionRect should not be called in ' +
+        'VisibilityManager for mApp', TAG);
+    return () => {};
+  }
+
+  /**
+   * @override
+   */
+  getRootLayoutBox() {
+    // By the time `#getRootLayoutBox` is called, it is garanteed that
+    // onVisibilityChangeHandler has been called at least once
+    return devAssert(this.intersectionRect_);
+  }
+
+  /**
+   * @param {!../../../src/inabox/host-services.VisibilityDataDef} visibilityData
+   * @private
+   */
+  onVisibilityChangeHandler_(visibilityData) {
+    if (this.disposed_) {
+      return;
+    }
+    //TODO: Need discussion
+    // rootVisibility is set by hostAPI, instead of Viewer.isVisible
+    let ratio = visibilityData.visibleRatio;
+    // Convert to valid ratio range in [0, 1]
+    ratio = ratio < 0 ? 0 : ratio;
+    ratio = ratio > 1 ? 1 : ratio;
+    this.setRootVisibility(ratio);
+    this.intersectionRect_ = visibilityData.visibleRect;
+  }
+
+  /**
+   * @override
+   */
+  observe() {
+    devAssert(false, '%s: element level visibility not supported, ' +
+        'getElementIntersectionRect should not be called in ' +
+        'VisibilityManager for mApp', TAG);
+    return () => {};
+  }
+
+  /**
+   * @override
+   */
+  getElementVisibility() {
+    devAssert(false, '%s: element level visibility not supported, ' +
```

Comment:
do you want to throw or log error ?

Reply:
I do want to throw. Supposedly VisibilityManagerForHostAPI 's `getElementVisibility()` should never be invoked. I implemented this simply because closure compiler forced me to.